rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -57,93 +69,245 @@ def _maybe_decode_v3_swap(  # pylint: disable=no-self-use
+        # transfer event (and other events) are parts of the swap.
+        if decoded_events[1].asset == A_ETH:
+            if decoded_events[1].event_type == HistoryEventType.SPEND:
+                # This is a swap from eth to token
+                swap_data = self._decode_eth_to_token_swap(decoded_events)
+
+            else:  # Receive
+                # This is a swap from token to eth
+                swap_data = self._decode_token_to_eth_swap(decoded_events)
+
+        else:
+            # Then this should be a swap from token to token
+            swap_data = self._decode_token_to_token_swap(decoded_events)
+
+        if swap_data is None or swap_data.from_asset is None or swap_data.to_asset is None:
+            self.msg_aggregator.add_warning(
+                f'Failed to decode a uniswap swap for transaction {transaction.tx_hash.hex()}. '
+                f'Please open an issue on our github or contact us in our discord server',
+            )
+            return decoded_events
+
+        try:
+            from_crypto_asset = swap_data.from_asset.resolve_to_crypto_asset()
+        except (UnknownAsset, WrongAssetType):
+            self.notify_user(event=decoded_events[1], counterparty=CPT_UNISWAP_V3)
+            return decoded_events
+
+        try:
+            to_crypto_asset = swap_data.to_asset.resolve_to_crypto_asset()
+        except (UnknownAsset, WrongAssetType):
+            self.notify_user(event=decoded_events[-1], counterparty=CPT_UNISWAP_V3)
+            return decoded_events
+
+        from_event = HistoryBaseEntry(
+            event_identifier=transaction.tx_hash,
+            sequence_index=decoded_events[1].sequence_index,
+            timestamp=decoded_events[1].timestamp,
+            location=Location.BLOCKCHAIN,
+            location_label=transaction.from_address,
+            asset=from_crypto_asset,
+            balance=Balance(amount=swap_data.from_amount),
+            notes=f'Swap {swap_data.from_amount} {from_crypto_asset.symbol} via uniswap auto router',  # noqa: E501
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.SPEND,
+            counterparty=CPT_UNISWAP_V3,
+        )
+
+        to_event = HistoryBaseEntry(
+            event_identifier=transaction.tx_hash,
+            sequence_index=decoded_events[-1].sequence_index,
+            timestamp=decoded_events[-1].timestamp,
+            location=Location.BLOCKCHAIN,
+            location_label=transaction.from_address,
+            asset=to_crypto_asset,
+            balance=Balance(amount=swap_data.to_amount),
+            notes=f'Receive {swap_data.to_amount} {to_crypto_asset.symbol} as the result of a swap via uniswap auto router',  # noqa: E501
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.RECEIVE,
+            counterparty=CPT_UNISWAP_V3,
+        )
+
+        gas_event = decoded_events[0]
+        return [gas_event, from_event, to_event]
```

Comment:
Sorry, didn't get the question.

> only 3 events in the original list of decoded events

Why does the length of the original list matters here?

Reply:
Here since you check the transaction.to_address I believe is fine. In the case of a transaction that bundles multiple events you might have issues but is a case that we don't handle yet so is fine